src/ssl/support.h

```diff
@@ -273,25 +294,28 @@ bool configureSSLUsingPkeyAndCertFromMemory(SSL *ssl, const char *data, AnyP::Po
  */
 void useSquidUntrusted(SSL_CTX *sslContext);
 
-/**
-   \ingroup ServerProtocolSSLAPI
-   * Iterates over the X509 common and alternate names and to see if  matches with given data
-   * using the check_func.
-   \param peer_cert  The X509 cert to check
-   \param check_data The data with which the X509 CNs compared
-   \param check_func The function used to match X509 CNs. The CN data passed as ASN1_STRING data
-   \return   1 if any of the certificate CN matches, 0 if none matches.
- */
-int matchX509CommonNames(X509 *peer_cert, void *check_data, int (*check_func)(void *check_data,  ASN1_STRING *cn_data));
```

Comment:
This C-style function was one of the primary sources of problems fixed in this PR. As you can see, it was using essentially two void* pointers (ASN1_STRING can be nearly anything) and expected the passed callback (i.e. check_func) to figure out what X.509 name it is being called with. The replacement GeneralNameMatcher API (declared below) provides strong visual clues for developers backed by compile-time type checks to make sure we are not trying to compare a domain name with an IP address or simply forget to handle an IP address case.